lightning/src/ln/channelmanager.rs

```diff
@@ -2514,7 +2514,19 @@ impl<ChanSigner: ChannelKeys, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>
 						}));
 					}
 				}
-				try_chan_entry!(self, chan.get_mut().update_add_htlc(&msg, pending_forward_info), channel_state, chan);
+				let temp_chan_failure_reason = match pending_forward_info {
+					PendingHTLCStatus::Forward(PendingHTLCInfo { incoming_shared_secret, .. }) => {
+						let chan_update = self.get_channel_update(chan.get()).unwrap();
+						Some(onion_utils::build_first_hop_failure_packet(&incoming_shared_secret, 0x1000|7, &{
+							let mut res = Vec::with_capacity(8 + 128);
+							res.extend_from_slice(&byte_utils::be16_to_array(chan_update.contents.flags));
+							res.extend_from_slice(&chan_update.encode_with_len()[..]);
+							res
+						}[..]))
+					},
+					_ => None,
+				};
+				try_chan_entry!(self, chan.get_mut().update_add_htlc(&msg, pending_forward_info, temp_chan_failure_reason), channel_state, chan);
```

Comment:
This is a little hard to follow. Are we preemptively building a `msgs::OnionErrorPacket` in case it is needed by `update_add_htlc`? If so, it might be worth refactoring the code so the behavior isn't obscured across methods. Left another comment in `update_add_htlc` that may be relevant.